Support creating non-upgradeable contract instances

facuspagnuolo opened the pull request:
Fixes https://github.com/zeppelinos/zos-lib/issues/32
Details explained here: https://github.com/zeppelinos/zos-cli/pull/318

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,26 @@
+
+  const params = Object.assign({}, contractClass.defaults(), txParams, { to: 0x0, data: ASM_CODE_COPY })
+  const txHash = await web3.eth.sendTransaction(params)
+  const receipt = await web3.eth.getTransactionReceipt(txHash)
```
I don't think that `getTransactionReceipt` blocks until the tx is mined, but rather it may return null if it is still pending. Can we check this manually on geth or parity? I'm also a bit worried as to how to detect these kind of errors when our entire test suite is on ganache.

spalladino replied:
Also: keep in mind that web3 does not return promises, so the await is unneeded, since the version without callback simply blocks the entire thread (we may want to use callbacks though, or use a promisified version of web3 as buidler does).

facuspagnuolo replied:
That's true, actually both functions accept a callback to work asynchronously. Let's do that to avoid an issue

spalladino commented on src/app/App.js:
```diff
@@ -103,6 +104,13 @@ export default class App {
     log.info(` Admin for proxy ${proxyAddress} set to ${newAdmin}`)
   }
 
+  async createNonUpgradeableInstance(contractClass, contractName, initMethodName, initArgs) {
```
How about `createContract` or `copyContract` instead? `nonUpgradeableInstance` does not sound very good, although I concur that it is very descriptive.

facuspagnuolo commented:
@spalladino addressed your comments. I tried to use `web3` working with callbacks but the `contractAddress` is never returned in that case after deploying the contract.

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,26 @@
+
+  const ASM_CODE_COPY = `0x73${address}803b8091600080913c6000f3`
+
+  const params = Object.assign({}, contractClass.defaults(), txParams, { to: 0x0, data: ASM_CODE_COPY })
```
It seems that sending a transaction to `0x0` (in geth and parity) does not actually create a new contract, but fails because it is a malformed address (ganache is happy to accept it). Sending to the `0x0....0` address is accepted in geth and parity, but actually sends a tx to that address (instead of creating a contract); ganache instead fails because it is not a valid contract address (same with any other address).

In order to actually create a contract (in all nodes), one needs to send the tx to `null` instead of `0x0`. Yet another point for testing on geth instead of ganache.

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,26 @@
+
+  const params = Object.assign({}, contractClass.defaults(), txParams, { to: 0x0, data: ASM_CODE_COPY })
+  const txHash = web3.eth.sendTransaction(params)
+  const receipt = web3.eth.getTransactionReceipt(txHash)
```
Web3's `getTransactionReceipt` blocking version will return once the request is sent to the node, **not** once the transaction is mined. This method will always return the receipt in ganache (since it mines almost instantly), but in geth or parity will return `null` until the tx is actually mined.

We actually need to add a loop that checks every X time for the tx to be mined, which is what truffle ends up doing for regular transactions ([see here](https://github.com/trufflesuite/truffle/blob/develop/packages/truffle-contract/contract.js#L140)), or use web3 `checkForContractAddress`, which does something similar but block-based instead of time-based ([see here](https://github.com/ethereum/web3.js/blob/develop/lib/web3/contract.js#L96)).

facuspagnuolo commented:
@spalladino

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,55 @@
+const RECEIPT_CHECK_TIMEBOX = 100
+const DEPLOYMENT_TIMEOUT = 60000
```
Could we use the value of the timeout flag for this?

facuspagnuolo replied:
sure

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,55 @@
+const RECEIPT_CHECK_TIMEBOX = 100
```
I'd increase this to at least 1s, 100ms seems too aggressive

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,55 @@
+      web3.eth.getTransactionReceipt(txHash, callback(resolve, reject))
+    })
+    if (receipt) return receipt
+    else await sleep(100)
```
Should be `RECEIPT_CHECK_TIMEBOX`, and not the hardcoded number

spalladino commented on src/helpers/copyContract.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+async function getTransactionReceipt(txHash) {
+  for (let i = 0; i < (DEPLOYMENT_TIMEOUT / RECEIPT_CHECK_TIMEBOX) ; i++) {
```
Note that this loop will end well past the deployment timeout, since it's not accounting for how long the body of the for-loop runs (and getTransactionReceipt may be quite lengthy). 

If we consider this to be an issue (it may very well not be), we can either:
- Use a lodash-debounced version of the function, where we call back into the function to iterate, and don't need to manually `sleep` on every iteration (extra points because lodash!)
- Run the for-loop every TIMEBOX ms, but abort after TIMEOUT ms have passed, and not after TIMEOUT/TIMEBOX iterations.

spalladino commented:
Also, after we merge this one and #211, we should estimate the gas for this call.

spalladino commented:
Awesome work :-)
